server/src/main/java/org/opensearch/common/bytes/AbstractBytesReference.java

```diff
@@ -82,7 +84,13 @@ public void writeTo(OutputStream os) throws IOException {
 
     @Override
     public String utf8ToString() {
-        return toBytesRef().utf8ToString();
+        BytesRef bytesRef = toBytesRef();
+        final char[] ref = new char[bytesRef.length];
+        final int len = UnicodeUtil.UTF8toUTF16(bytesRef.bytes, bytesRef.offset, bytesRef.length, ref);
+        if (len > MAX_LENGTH) {
+            throw new IllegalStateException("UTF16 String size is " + len + ", should be less than " + MAX_LENGTH);
```

Comment:
1. Should this throw an `IllegalStateException` or `IllegalArgumentException`? Was `IllegalStateException` chosen for a specific reason?
2. Please could you help add a unit test that validates that an `OutOfMemoryException` is not thrown now and instead the new exception type (which shouldn't cause a node drop) is thrown?

Reply:
It should be `NotSupportedException`

Reply:
Hm ... which `NotSupportedException` you mean?

Reply:
> It should be NotSupportedException

@gashutos which one do you suggest? There isn't any `NotSupportedException`. Maybe we can go with `IllegalArgumentException`?

Reply:
I like that.